crates/dekaf/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+use anyhow::Context;
+use bytes::BufMut;
+use kafka_protocol::{
+    messages,
+    protocol::{Builder, Decodable, Encodable},
+};
+
+mod topology;
+use topology::{fetch_all_collection_names, Collection, Partition};
+
+mod read;
+use read::{read_record_batch, Read};
+
+mod session;
+pub use session::Session;
+
+pub mod registry;
+
+pub struct App {
+    /// Anonymous API client for the Estuary control plane.
+    pub anon_client: postgrest::Postgrest,
+    /// Hostname which is advertised for Kafka access.
+    pub advertise_host: String,
+    /// Port which is advertised for Kafka access.
+    pub advertise_kafka_port: u16,
+}
+
+/// Dispatch a read request `frame` of the current session, writing its response into `out`.
+/// `raw_sasl_auth` is the state of SASL "raw" mode authentication,
+/// and conditions the interpretation of request frames.
+#[tracing::instrument(level = "debug", err(level = "warn"), skip_all)]
+pub async fn dispatch_request_frame(
+    session: &mut Session,
+    raw_sasl_auth: &mut bool,
+    frame: bytes::BytesMut,
+    out: &mut bytes::BytesMut,
+) -> anyhow::Result<()> {
+    use messages::*;
+
+    /*
+    println!(
+        "full frame:\n{}",
+        hexdump::hexdump_iter(&frame)
+            .map(|line| format!(" {line}"))
+            .collect::<Vec<_>>()
+            .join("\n")
+    );
+    */
+
+    let (api_key, version) = if !*raw_sasl_auth {
```

Comment:
TIL that the kafka protocol uses the term "apiKey" to mean _not_ an authentication token 🙃

Reply:
This experience did nothing to improve my opinion of the kafka wire API.